lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,378 @@
+
+        return height_diff_mask
+
+    def process(self, temperature_cube, orography_cube, land_sea_mask_cube):
+        """Calculates the lapse rate from the temperature and orography cubes.
+
+        Args:
+            temperature_cube (iris.cube.Cube):
+                Cube of air temperatures (K).
+
+            orography_cube (Iris.cube.Cube):
+                Cube containing orography data (metres)
+
+            land_sea_mask_cube (iris.cube.Cube):
+                Cube containing a binary land-sea mask. True for land-points
+                and False for Sea.
+
+        Returns:
+            lapse_rate_cube (iris.cube.Cube):
+                Cube containing lapse rate (Km-1)
+
+        Raises
+        ------
+        TypeError: If input cubes are not cubes
+        ValueError: If input cubes are the wrong units.
+
+        """
+
+        if not isinstance(temperature_cube, iris.cube.Cube):
+            msg = "Temperature input is not a cube, but {}"
+            raise TypeError(msg.format(type(temperature_cube)))
+
+        if not isinstance(orography_cube, iris.cube.Cube):
+            msg = "Orography input is not a cube, but {}"
+            raise TypeError(msg.format(type(orography_cube)))
+
+        if not isinstance(land_sea_mask_cube, iris.cube.Cube):
+            msg = "Land/Sea mask input is not a cube, but {}"
+            raise TypeError(msg.format(type(land_sea_mask_cube)))
+
+        # Converts cube units.
+        temperature_cube.convert_units('K')
+        orography_cube.convert_units('metres')
+
+        enforce_float32_precision([temperature_cube])
+
+        # Extract x/y co-ordinates.
+        x_coord = temperature_cube.coord(axis='x').name()
+        y_coord = temperature_cube.coord(axis='y').name()
+
+        # Extract orography and land/sea mask data.
+        orography_data = next(orography_cube.slices([y_coord,
+                                                     x_coord])).data
+        land_sea_mask = next(land_sea_mask_cube.slices([y_coord,
+                                                        x_coord])).data
+        # Fill sea points with NaN values.
+        orography_data = np.where(land_sea_mask, orography_data, np.nan)
+
+        # Extract data array dimensions to define output arrays.
+        dataarray_shape = next(temperature_cube.slices([y_coord,
+                                                        x_coord])).shape
+        dataarray_size = dataarray_shape[0] * dataarray_shape[1]
+
+        # Array containing all of the subsections extracted from data array.
+        # Also enforce single precision to speed up calculations.
+        all_temp_subsections = np.zeros((dataarray_size,
+                                         self.nbhoodarray_size),
+                                        dtype=np.float32)
+        all_orog_subsections = np.zeros((dataarray_size,
+                                         self.nbhoodarray_size),
+                                        dtype=np.float32)
```

Comment:
Point of preference: the indentation here is technically correct but looks a little weird.  I think we might more often (for IMPROVER) go with a line break after the first bracket, eg:
```
    all_temp_subsections = np.zeros(
        (dataarray_size, self.nbhoodarray_size), dtype=np.float32)
```
So would you mind tweaking that please?